packages/files-s3/src/drivers/s3.ts

```diff
@@ -0,0 +1,374 @@
+        isFile: true,
+        etag: stats.ETag
+      }
+    } catch (error) {
+      throw new CannotGetMetaDataException(location, 'stats', error)
+    }
+  }
+
+  /**
+   * Returns the signed url for a given path
+   */
+  public async getSignedUrl(location: string, options?: SignedUrlOptions): Promise<string> {
+    try {
+      return await getSignedUrl(
+        this.adapter,
+        new GetObjectCommand({
+          Key: location,
+          Bucket: this._config.bucket,
+          ...this._transformContentHeaders(options)
+        }),
+        {
+          expiresIn: options?.expiresIn ?? 15 * 60
+        }
+      )
+    } catch (error) {
+      throw new CannotGetMetaDataException(location, 'signedUrl', error)
+    }
+  }
+
+  /**
+   * Returns URL to a given path
+   */
+  public async getUrl(location: string): Promise<string> {
+    /**
+     * Use the CDN URL if defined
+     */
+    if (this._config.cdnUrl) {
+      return `${this._config.cdnUrl}/${location}`
+    }
+
+    const href = format(await this.adapter.config.endpoint())
+    if (href.startsWith('https://s3.amazonaws')) {
+      return `https://${this._config.bucket}.s3.amazonaws.com/${location}`
+    }
+
+    return `${href}/${this._config.bucket}/${location}`
+  }
+
+  /**
+   * Write string|buffer contents to a destination. The missing
+   * intermediate directories will be created (if required).
+   */
+  public async put(location: string, contents: Buffer | string, options?: WriteOptions): Promise<void> {
+    try {
+      await this.adapter.send(
+        new PutObjectCommand({
+          Key: location,
+          Body: contents,
+          Bucket: this._config.bucket,
+          ...this._transformWriteOptions(options)
+        })
+      )
+    } catch (error) {
+      throw new CannotWriteFileException(location, error)
+    }
+  }
+
+  /**
+   * Write a stream to a destination. The missing intermediate
+   * directories will be created (if required).
+   */
+  public async putStream(
+    location: string,
+    contents: NodeJS.ReadableStream,
+    options?: WriteOptions & {
+      multipart?: boolean
+      queueSize?: number
+      partSize?: number
+      leavePartsOnError?: boolean
+      tags?: Tag[]
+      tap?: (stream: Upload) => void
+    }
+  ): Promise<void> {
+    try {
+      options = Object.assign({}, options)
+
+      /**
+       * Upload as multipart stream
+       */
+      if (options.multipart) {
+        const { tap, queueSize, partSize, leavePartsOnError, tags, ...others } = options
+        const upload = new Upload({
+          params: {
+            Key: location,
+            Body: contents,
+            Bucket: this._config.bucket,
+            ...this._transformWriteOptions(others)
+          },
+          queueSize,
+          partSize,
+          leavePartsOnError,
+          tags,
+          client: this.adapter
+        })
+
+        if (typeof tap === 'function') {
+          tap(upload)
+        }
+
+        await upload.done()
```

Comment:
I think you should not await for the upload to be done here, the user can invoke your method to start streaming to S3 and probably want to do something else.

Instead you should return the upload.done Promise.
This way, people using your lib can await themselves when they need to.

You could also return the whole upload object, so people can listen to `httpProgressEvent` and so on.
But this means you would have to introduce a type generic in your DriverContract as every drivers implementation would return a different stuff.

Generally speaking for this kind of stream helpers, I like to return a WritableStream, so you can easily pipe it to other streams e.g. :

```typescript

export const uploadStream = async (Bucket: string, Key: string): Promise<{ uploadStream: Writable, uploadController: Upload}> => {
  const pass = new PassThrough();
  const upload = new Upload({
    client: new S3Client(),
    params: { Bucket , Key, Body: pass }
  })
  return {
    uploadStream: pass,
    uploadController: upload,
  };
};

// Call the helper
const { uploadStream, uploadController } = await uploadStream('my-bucket', 'huge-archive.tar.gz');

// Pipe other streams in the exposed WriteStream
// for instance 
createReadStream('path/to/my-huge-archive.tar.gz').pipe(uploadStream)
// or more complex things such as
tarStream.pipe(gunzipStream).pipe(uploadStream)

// then you can something else, the upload is started
doOtherThings();
// you can also do whatever you want with progress event
upload.on('httpUploadProgress', console.info);

// when you need to be sure upload is done, just await for it
await upload.done()
```
Does that make sense ?